### ISSUE TITLE
Support .NET Core tool installation

### ISSUE THREAD

**Issue by jfro** [NetSparkleUpdater/NetSparkle#115]
Thought an easy way to get the NetSparkle.Tools installed and available for me to use might be using `dotnet tool install` but it looks like it's not a compatible package type to do so. Wondering if this might be a useful thing to do, to make it that type of package?

Trying gets this which troubleshooting says it's a regular nuget package instead of a Core Tool package:
```
NU1212: Invalid project-package combination for <ToolName>. DotnetToolReference project style can only contain references of the DotnetTool type.
```

**Comment by Deadpikle** [comment]
This is a great idea. Thank you for the suggestion.

Note to self/whomever: https://docs.microsoft.com/en-us/dotnet/core/tools/global-tools-how-to-create is a tutorial on how to do this. Will need to figure out how to update the current NuGet and/or the CI/CD to do all this.

**Comment by aelmendorf** [comment]
I downloaded source then just added below 3 lines to project xml
`<PackAsTool>true</PackAsTool>`
`<ToolCommandName>appcast</ToolCommandName>`
` <PackageOutputPath>./nupkg</PackageOutputPath>`

then it was a simple as `dotnet tool install --global --add-source ./nupkg generate_appcast`

Im not sure  if this would let you install without unpacking nuget from web before trying to install or not.  I think user would need to download, unpack then install using above dotnet command

**Comment by johncclayton** [comment]
first time center, but I've been lurking a while.

what would it require to have the generator tool made available as a NuGet package as part of the normal release?

I'm adding something like tests, a release process of some kind and then something like github actions to bring this together.

I'm willing to help agent this available (I've just put NetSparkle into an app i ported from mac), and having the tool available on whatever I'm using to rebuild the appcast would be very helpful.

if it's already been done and I've not seen it - apologies in advance.

**Comment by Deadpikle** [comment]
@johncclayton Thanks for your question. Does https://www.nuget.org/packages/NetSparkleUpdater.Tools/ help? It should have the app cast generator in it.

**Comment by johncclayton** [comment]
Absolutely - I checked the code out to work out what I was doing wrong on the command line.  I've got a lot of experience in IT (over 20 years), but am really new to the whole .net / CLI part - so while the tools worked, I didn't know how to get the generator .exe tool installed reliably at a path location I could use for the build.

But if it's just something like: dotnet install that I have to do - it sounds done already :-) 

When I installed the package the exe ended up in C:\Users\john\.nuget\packages\netsparkleupdater.tools\2.0.0-preview20210131001 - and of course this isn't on a standard build path.  If there's a way to get that tool available in a GitHub action more easily - I'd love to know about it.

Thanks for the fast answer.

**Comment by johncclayton** [comment]
I tried the changes as outlined by @aelmendorf - that worked really great - thank you.  It probably wants to be called something different than appcast, likely generate_appcast.exe right?  I can create a PR from my own forked version of the repo if that'd help.

**Comment by Deadpikle** [comment]
Hi, sorry, not wanting to ignore you. Just very busy. 

I have two pending questions/things to research in my mind about this that need to be figured out:

1. How do I properly publish this to NuGet? I probably have to have 1 tool per 1 NuGet package, I'd assume, rather than the current setup of all tools in 1 package?
2. Is everything setup properly for the tool to list things on all 3 platforms such that `dotnet tool install` will work right? I have no clue if I've set up the NuGet package properly.

As far as a name: `netsparkle-generate-appcast` is what I would prefer, personally, to differentiate it with the macOS Sparkle library.

tl;dr: I have research to do to understand how all of this works and if the CI/CD and NuGet need restructuring. I won't have time to do this for a while. If someone else **knows** how this works, it'd be greatly appreciated.

**Comment by johncclayton** [comment]
I get busy, no problem at all.  I am more than happy to start looking for answers to these questions - as part of my learning journey into .net. 

I believe at a minimum, having the generate-app-cast tool in a separate NuGet package makes a TON of sense.  I like the cross platform aspect of that a LOT, and I have the gear here to test it on all 3 platforms if that helps.   I don't have an M1 thing, but I know someone who does. 

Got any good links to help me get up to speed on NuGet packaging / building them / best practises?  I'm totally willing to learn this to help out.

**Comment by johncclayton** [comment]
Ok, tomorrow I'll read the following properly:

    1. https://docs.microsoft.com/en-us/dotnet/core/tools/global-tools-how-to-create 
    2. https://docs.microsoft.com/en-us/dotnet/core/tools/local-tools-how-to-use
    
I think I'll play around with using a manifest file and see if I can get it to a stage where I'm able to change my build process to simply do: 

    $ dotnet tool run netsparkle-generate-appcast <arguments>

Please forgive me ahead of time if I'm spouting utter rubbish, the tooling and how it all hangs together (and what its possibilities and limitations are) is still really new to me. 

Sleep calls.
(2am).

**Comment by johncclayton** [comment]
Hi

I spend a couple of hours learning / experimenting with the nuget tool & packaging.  I learned the ropes by creating a simple tool project, and published this locally.  I then proceeded to install it from my NAS on Mac, Windows and Linux systems.

The installation works like a charm in both global and local.  The tool .exe/command can be called whatever you like (and dotnet tool upgrade takes care of tool name changes across versions - which is a nice touch). 

I tested the behaviour on Windows, Linux (via VM) and Mac - all of them seem to work exactly the same way.  No surprise there really; hats off to the dotnet 5.0 dev work that was done to bring the same codebase to each of those platforms!

In my case I'm interested in having this tool available without having to mess with the PATH; doubly so when it comes to cross platform builds.  For this use case I would choose a local install as it allows me to run the command via `dotnet tool run <name-of-tool>`, so I can use the same command across all platforms.  I find that valuable as its less stuff that I need to keep in my head "just because I'm on Mac and not Windows etc etc".  Additionally if I switch branches and require some other version of the tool this local install can be changed without breaking other projects.  

Here's what I'd suggest as next steps: 
1. update the documentation for newbies (like me); such that it describes explicitly how to set up a working build environment & the right way to install the tool - this is important; because there is a "wrong" way as well (by trying to use the .Tools package as a dependency - I'll explain that below). 
2. I'd like to write up the stuff I would have *loved* to have the first time I saw the docs - I came into NetSparkle with no nuget tool knowledge, and didn't know about local vs global - having a bit of background will certainly help other devs reach success and also help NetSparkle gain even more traction. 
3. Change the csproj XML tags for the tool as suggested by @aelmendorf - additionally using the ToolCommandName tag as you wanted the thing to have a very specific name.
````
     <PackAsTool>true</PackAsTool>
     <ToolCommandName>netsparkle-generate-appcast</ToolCommandName>
````

4. BTW: I don't see a need to separate out the tools into a separate NuGet package, that just seems like more work for no benefit.

Doing this allows the tools to be used something like this - anywhere - e.g. GitHub Actions etc - note the use of a local tool to make it possible to run it without having to know the path name:

    $ dotnet new tool-manifest
    $ dotnet install tool NetSparkle.Tools (a local install - not global)
    $ dotnet tool run netsparkle-generate-appcast <args>

After 3 hours of learning and playing around - I like that this patch appears to be simply two lines of XML configuration.  My goodness; if that's actually TRUE I'll be a happy man. 

Let me know what you think; I'd be happy to adjust the outstanding PR to include documentation changes as well. 

Thank you!

**Comment by Deadpikle** [comment]
@johncclayton Thank you SO MUCH for your contributions, here. That's very helpful to confirm that the csproj changes are really all that is needed!

I have done the following:
* Merged your PR with those changes
* Published two NEW packages: `NetSparkleUpdater.Tools.AppCastGenerator` and `NetSparkleUpdater.Tools.DSAHelper`. Two new packages keeps things separate (versions, etc.) and also makes it possible to allow BOTH of these tools to work as .NET global tools.
  * `dotnet tool install --global NetSparkleUpdater.Tools.AppCastGenerator` -> `netsparkle-generate-appcast`
  * `dotnet tool install --global NetSparkleUpdater.Tools.DSAHelper` -> `netsparkle-dsa`
* Bumped tool versions to .NET 5 instead of .NET Core 3.1
* Deprecated `NetSparkleUpdater.Tools` in lieu of the two new packages (plus some issues with having that as one package when there is `PackAsTool` set -- see [this GitHub issue in dotnet](https://github.com/dotnet/sdk/issues/10053))

> update the documentation for newbies (like me); such that it describes explicitly how to set up a working build environment & the right way to install the tool - this is important; because there is a "wrong" way as well (by trying to use the .Tools package as a dependency - I'll explain that below).

Please feel free to submit a PR! You can make a new section in the README for this.

> I'd like to write up the stuff I would have loved to have the first time I saw the docs - I came into NetSparkle with no nuget tool knowledge, and didn't know about local vs global - having a bit of background will certainly help other devs reach success and also help NetSparkle gain even more traction.

Again, please feel free to submit a PR!

I'm going to go ahead and close this issue as "finished" -- for doc update requests, please comment on #61 or create a new issue for that specific documentation.

**Thank you** for your patience and for everyone's help!
